### PR TITLE
Global shuffle, 1x egress

### PR DESCRIPTION
**Pull request by knighton**

### CODE REVIEW

File: composer/datasets/streaming/dataset.py [hunk 1]
```diff
@@ -1,328 +1,467 @@
 # Copyright 2022 MosaicML Composer authors
 # SPDX-License-Identifier: Apache-2.0
 
-"""The :class:`StreamingDataset` class, used for building streaming iterable datasets.
```
Comment (ravi-mosaicml) [inline composer/datasets/streaming/dataset.py:4]: Can you leave the class level docstring?

**Review by ravi-mosaicml — COMMENTED** [review]
I didn't give this PR a close review, but at a high level, I am wondering if we can avoid setting up our own sockets. 

Instead, could we use `torch.dist` for all main-process to main-process communication, and then use `multiprocessing` for all main-process-to-dataloader-worker communication? Not a huge fan of setting up our own sockets when pytorch already provides (optimized) primitives to help with this.

**Comment by ravi-mosaicml** [comment]
Decided against this -- won't be sharing shards across nodes.
